### PR TITLE
Added codeblock support with ` char.

### PR DESCRIPTION
**Pull request by anoadragon453**
Changes font to monospace with a gray color. Wanted to do a different background as well but requires QString be embedded in one of Qt's [rich text processing APIs](https://doc.qt.io/qt-5/richtext.html). 

Markdown inside codeblocks is not triggered.

Addresses #2999.

### CODE REVIEW

**Comment by anoadragon453** [comment]
This however does not protect against long lines of code triggering a message split, as mentioned [here](https://github.com/tux3/qTox/issues/2999#issuecomment-195777960).

**Comment by zetok** [comment]
line

Doesn't work when line ends with a `\n`. Other modes are affected by it as well.

**Comment by anoadragon453** [comment]
Thanks for pointing that out, md checks for either an end-of-line or space to be parsed properly, if there's an `\n` being inserted, it doesn't match either of those criteria. I'll try to patch those up for all modes.

**Comment by anoadragon453** [comment]
@zetok Newline issue fixed, I did notice that the parsing does accidentally succeed if we type something like:

```
*vv*'
```

when really it should fail as the apostrophe is there. 

Essentially it fails if a character is not present to the left of the markdown, while something _is_ present to the right.

File: src/chatlog/chatmessage.cpp [hunk 1]
```diff
@@ -209,8 +210,10 @@ QString ChatMessage::detectMarkdown(const QString &str)
 
             QString htmledSnippet;
 
-            // Check for surrounding spaces and/or beginning/end of line
-            if (!((snipCheck.startsWith(' ') || offset == 0) && (snipCheck.endsWith(' ') || offset + snipCheck.trimmed().length() == out.length())))
+            qDebug() << "snip: " << snipCheck << " off+leng: " << offset + snipCheck.trimmed().length() << " out.len: " << out.length() << " offset: " << offset;
```
Comment (zetok) [inline src/chatlog/chatmessage.cpp]: This debug statement isn't needed any longer

File: src/chatlog/chatmessage.cpp [hunk 2]
```diff
@@ -209,8 +210,10 @@ QString ChatMessage::detectMarkdown(const QString &str)
...
+
+            // Check for surrounding spaces, newlines and/or beginning/end of line
+            if (!(((snipCheck.startsWith(' ') || snipCheck.startsWith('>')) || offset == 0) || ((snipCheck.endsWith(' ') || snipCheck.endsWith('<')) && offset + snipCheck.trimmed().length() == out.length())))
```
Comment (zetok) [inline src/chatlog/chatmessage.cpp]: I wonder if that could be done differently

File: src/chatlog/chatmessage.cpp (hunk 1 above)
Comment (anoadragon453) [inline src/chatlog/chatmessage.cpp]: Good catch, I'll remove it.

File: src/chatlog/chatmessage.cpp (hunk 2 above)
Comment (anoadragon453) [inline src/chatlog/chatmessage.cpp]: It is quite confusing and scary, no?

I'll look into revising it.

Comment (anoadragon453) [inline src/chatlog/chatmessage.cpp]: After trying to convert this to a singular QRegExp for the past three hours I can say this is the easier route.

**Comment by sudden6** [comment]
@anoadragon453 can you fix this so your case from above doesn't parse as md?

**Comment by anoadragon453** [comment]
@sudden6 Will get on that now.

File: src/chatlog/chatmessage.cpp [hunk 3]
```diff
@@ -209,38 +210,39 @@ QString ChatMessage::detectMarkdown(const QString &str)
...
-            // Check for surrounding spaces and/or beginning/end of line
-            if (!((snipCheck.startsWith(' ') || offset == 0) && (snipCheck.endsWith(' ') || offset + snipCheck.trimmed().length() == out.length())))
+            qDebug() << snipCheck.toHtmlEscaped() << " offset: " << offset << "snipLength: " << snippet.toHtmlEscaped().trimmed().length() << " out.length: " << out.length() << " out: " << out;
+
+            // Only parse if surrounded by spaces, newline(s) and/or beginning/end of line
+            if ((snipCheck.startsWith(' ') || snipCheck.startsWith('>') || offset == 0) && ((snipCheck.endsWith(' ') || snipCheck.endsWith('<')) || offset + snippet.toHtmlEscaped().length() == out.length()))
             {
+                int mul = 0; // Determines how many characters to strip from markdown text
+                // Set mul depending on markdownPreference
+                if (Settings::getInstance().getMarkdownPreference() == 2)
```
Comment (sudden6) [inline src/chatlog/chatmessage.cpp]: please replace this magic number

Comment (anoadragon453) [inline src/chatlog/chatmessage.cpp]: Thanks, fixed.

**Comment by anoadragon453** [comment]
All good to go :+1:

**Comment by anoadragon453** [comment]
_Now_, all good to go for merge :+1:

**Comment by ProMcTagonist** [comment]
It doesn't seem to be preserving whitespace.

nonpreserved whitespace

New issue or broken on my end?

**Comment by zetok** [comment]
@ProMcTagonist well, e.g. gh markdown also doesn't seem to preserve whitespaces, although in a slightly different manner

` t e s t  ` → `t e s t`

Although in a mode without stripping it probably shouldn't strip them.. so new issue it is.

**Comment by bignaux** [comment]
Confirmed bug here.
Interesting stuff :+1: 
https://doc-snapshots.qt.io/qt5-5.6/qtwebengine-webenginewidgets-markdowneditor-example.html
It would be neat to have preview in chatlog during typing.

**Comment by anoadragon453** [comment]
@bignaux That would be nice. Not exactly sure where we'd put it in the current chat view. Move the chatlog up and display it just above the input box?

**Comment by bignaux** [comment]
Yes, i was thinking to the same location as when you decide to press enter... with possibly the typing icon at right.

**Comment by anoadragon453** [comment]
Would it automatically come up upon detecting markdown or would it be manually invoked? We already have two buttons next to send.

**Comment by ProMcTagonist** [comment]
If there's a preview, it should appear above the textarea and update automatically.

But is a preview necessary? That leans toward a full formatting bar, and more clutter.

KISS. I'm against a preview.

**Comment by anoadragon453** [comment]
It's easy enough to set a bit when markdown is successfully detected.

**Comment by sudden6** [comment]
I'm with @ProMcTagonist I also don't really like the idea of a preview.
@anoadragon453 A new issue about not preserving whitespace is probably appropriate.

**Comment by zetok** [comment]
@sudden6 it was fixed in #2994.

**Comment by sudden6** [comment]
I thought this was about something else?

**Comment by zetok** [comment]
> I thought this was about something else?
> 
> @anoadragon453 A new issue about not preserving whitespace is probably appropriate.

Um, #2994 was about fixing trimming of whitespaces.

**Pull request merged**